### PR TITLE
Add payment state information to customer's installment listing

### PR DESCRIPTION
**Pull request by voneiden**
Relates to backend PR's City-of-Helsinki/apartment-application-service/pull/329 and City-of-Helsinki/apartment-application-service/pull/332

Demo of various randomly selected states:
image

Underlined text can be clicked to open a dialog showing individual payment dates and amounts
image

Normally if an installment has not yet been sent to SAP we'll just show "Unsent" instead of "Unpaid"
image

### CODE REVIEW

File: src/components/installments/InstallmentPaymentStateInfo.tsx [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+type StatusLevel = 'unimportant' | 'success' | 'important' | 'critical';
+
+const statusLevel = function statusLevel(state: ApartmentInstallmentPaymentState): StatusLevel {
+  if (state.status === 'PAID') {
+    if (state.overdue) {
+      return 'important';
+    }
+    return 'success';
+  } else if (state.overdue || state.status === 'OVERPAID') {
```
Comment (juhakujala) [inline src/components/installments/InstallmentPaymentStateInfo.tsx:36]: Same enum fix with these `OVERPAID` and `UNPAID`

File: src/components/installments/InstallmentPaymentStateInfo.tsx [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+const T_PATH = 'components.installments.InstallmentPaymentStateInfo';
+
+interface IProps {
+  state: ApartmentInstallmentPaymentState;
+  payments: ApartmentInstallmentPayment[];
+  sentToSap: boolean;
+}
+
+type StatusLevel = 'unimportant' | 'success' | 'important' | 'critical';
```
Comment (juhakujala) [inline src/components/installments/InstallmentPaymentStateInfo.tsx:28]: I would define these as either enums or const, as they are used in a few places in this file. As a const:

```
const StatusLevel = {
  UNIMPORTANT: "unimportant",
  SUCCESS: "success",
  IMPORTANT: "important",
  CRITICAL: "critical",
} as const;
```
Then the type would be:
```
type StatusLevelValue = typeof StatusLevel[keyof typeof StatusLevel];
```
And usage:
```
StatusLevel.CRITICAL
```

File: src/components/installments/InstallmentPaymentStateInfo.tsx [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        payment_date: formatDateTime(p.payment_date, true),
+      }))
+    : [];
+
+  return (
+    <>
+      <div
+        className={classNames(styles.container, styles[`status-${level}`], showPayments && styles.clickable)}
+        onClick={open}
+      >
```
Comment (juhakujala) [inline src/components/installments/InstallmentPaymentStateInfo.tsx:101]: It's not a good idea to use plain `<div>` as a clickable element without having defined it as a button and being able to focus it with a keyboard. I would render the wrapping element based on the `showPayments` value.

If there's something to open, then render an anchor element or a button. And if not, then just a plain `<div>`

File: src/components/installments/InstallmentPaymentStateInfo.tsx [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                { key: 'amount', headerName: t('amount') },
+                { key: 'payment_date', headerName: t('payment_date') },
+              ]}
+              rows={mappedPayments}
+              indexKey="index"
+              renderIndexCol={false}
+            />
+          </Dialog.Content>
+          <Dialog.ActionButtons>
+            <Button onClick={close}>Close</Button>
```
Comment (juhakujala) [inline src/components/installments/InstallmentPaymentStateInfo.tsx:135]: "Close" should be translated

File: src/components/installments/InstallmentPaymentStateInfo.tsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        </div>
+        {statusIcon(level)}
+      </div>
+      {showPayments && (
+        <Dialog
+          id="info-dialog"
+          aria-labelledby="payment-state-info-dialog-header"
+          isOpen={isOpen}
+          close={close}
+          closeButtonLabelText="Close info dialog"
```
Comment (juhakujala) [inline src/components/installments/InstallmentPaymentStateInfo.tsx:115]: "Close info dialog" should be translated

File: src/components/installments/InstallmentPaymentStateInfo.tsx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+interface IProps {
+  state: ApartmentInstallmentPaymentState;
+  payments: ApartmentInstallmentPayment[];
+  sentToSap: boolean;
+}
+
+type StatusLevel = 'unimportant' | 'success' | 'important' | 'critical';
+
+const statusLevel = function statusLevel(state: ApartmentInstallmentPaymentState): StatusLevel {
```
[the remark this answers is not in the document]
Comment (voneiden) [reply src/components/installments/InstallmentPaymentStateInfo.tsx:30]: If you mean using `=>` then it's mainly a personal preference to not use top level anonymous functions. But OK, it does indeed go against the code base convention, I'll change it.

[the remark this answers is not in the document]
Comment (juhakujala) [reply src/components/installments/InstallmentPaymentStateInfo.tsx:30]: Ah yes, my example was missing the `=>`

**Pull request merged**